File: get_links.js

const app = Application.currentApplication();
app.includeStandardAdditions = true;

//const playlist_id = app.displayDialog('Playlist id', { defaultAnswer: ''});
playlist_id = 'PLMf7voIq6C7ri7OahDXYfYdyzIeWS-w54'
//const keywordsFilepath = app.displayDialog('Keywords list file', { defaultAnswer: ''});
keywordsFilepath = '/Users/azvasquez/keywords';


// Getting the ytdn api key
ObjC.import('stdlib')
const apiKey = $.getenv('YTDN_API_KEY')

function writeFile(filepath, text){
  const fh = app.openForAccess(Path(filepath), {writePermission: true});
  app.setEof(fh, { to: 0})
  app.write(text, { to: fh, startingAt: app.getEof(fh)});
  app.closeAccess(fh);
}


function sendQuery(page_token){
  let items = []
  const playlist_url = `https://www.youtube.com/playlist?list=${playlist_id}`;
  const part = 'id,snippet'

  let query = `https://www.googleapis.com/youtube/v3/playlistItems?part=${part}&playlistId=${playlist_id}&key=${apiKey}`
  //const query = `https://www.googleapis.com/youtube/v3/playlistItems?part=${part}&playlistId=${playlist_id}&key=$YTDN_API_KEY`

  if (page_token)
      query += `&pageToken=${page_token}`


  const data = app.doShellScript(`curl "${query}"`);

  const responseData = JSON.parse(data)

  items = responseData.items
  console.log(items)

  if (responseData.nextPageToken){
    items.push(...sendQuery(responseData.nextPageToken));
  }
  return items;
}



const items = sendQuery(null)





function yt_url(video_id){
  return `https://youtu.be/${video_id}`;
}


const title_url_desc_tuples = items.map(item => [item.snippet.title, yt_url(item.snippet.resourceId.videoId), item.snippet.description])




function sanitize(s){
  return s.replace(/[^\x20-\x7E]+/g, '');
}


const keywordsList = app.read(Path(keywordsFilepath)).split('\n');
let usedItems = 'Features:\n';
const urls = []


title_url_desc_tuples.forEach(tuple => {
  let title;
  let url;
  let description;
  [title, url, description] = tuple;


  title = sanitize(title);
  const descHead = sanitize(description.split(' ').slice(0,10).join(' '));
  console.log(descHead);

  const correspondentName = descHead.replace(/.*[cC]orrespondent,? ([\w-]+ [\w-]+),? .*/, '$1');
  const bureauName = descHead.replace(/.*EBC (.*?) (Bureau )?[cC]orrespondent.*/, '$1') + ' Bureau';

  const keywordMatches = keywordsList.filter(keyword => description.includes(keyword) || title.includes(keyword));
  if (keywordMatches) {
    usedItems += ` - ${title} (${correspondentName}, ${bureauName}) ${url}\n`;
    urls.push(url);
  }
});




writeFile('./used_items', usedItems);

writeFile('./used_links', urls.join('\n'));

File: get_links.py
import requests
import json
import re
import os


import sys


playlist_id = sys.argv[2]
yt_api_key = os.environ.get('YTDN_API_KEY')

def build_query(page_token=None):

    global playlist_id, yt_api_key


    playlist_url = f'https://www.youtube.com/playlist?list={playlist_id}'
    part = 'id,snippet'
    query = f'https://www.googleapis.com/youtube/v3/playlistItems?part={part}&playlistId={playlist_id}&key={yt_api_key}'
    if page_token:
        query += f'&pageToken={page_token}'

    print(query)
    return query

query = build_query()
items = []


# Build items list
while True:
    response_data = requests.get(query).content
    response = json.loads(response_data)
    items.extend(response['items'])
    if not 'nextPageToken' in response:
        break
    query = build_query(response['nextPageToken'])



def yt_url(video_id):
    return f'https://youtu.be/{video_id}'

title_url_desc_tuples = [(title,yt_url(_id), desc)  for (title,_id, desc) in map(lambda i: (i['snippet']['title'], i['snippet']['resourceId']['videoId'], i['snippet']['description']), items)]


def sanitize(s):
    return s.encode('ascii', 'ignore').decode()

with open(sys.argv[1], 'r') as f:
    keywords = f.read().split('\n')
    print(keywords)




used_items = 'Featuring:\n'
urls = []

for title, url, desc in title_url_desc_tuples:

    title = sanitize(title)
    desc_head = sanitize(' '.join(desc.split()[:30]))

    correspondent_name = re.sub('.*[cC]orrespondent,? ([\\w-]+ [\\w-]+),? .*', '\\1',  desc_head)
    bureau_name = re.sub('.*EBC (.*?) (Bureau ?)[cC]orrespondent.*', '\\1',  desc_head)

    for keyword in keywords:
        if (keyword.lower() in title.lower() or keyword.lower() in desc_head.lower()):
            break
    else:
        continue

    used_items += f' - {title} ({correspondent_name}, {bureau_name}) {url}\n'
    urls.append(url)


#used_items = '\n'.join(used_items_lines)
with open('used_items', 'w') as f:
    f.write(used_items)

with open('used_links', 'w') as f:
    f.write('\n'.join(urls))

